File: mobile-sim/python/mobilesim/rosie/perception/Robot.py
import math

from pysoarlib import *


VIEW_DIST = 8.0
VIEW_ANGLE = math.pi/2 * .6
VIEW_HEIGHT = 2.0

# Returns vertices for a 3D triangular region representing the view area
def get_view_region_verts():
    verts = []

    dx = VIEW_DIST/2
    dy = VIEW_DIST * math.sin(VIEW_ANGLE/2)
    dz = VIEW_HEIGHT/2
    # Top triangle
    verts.append("%f %f %f" % (-dx, 0.0, dz))
    verts.append("%f %f %f" % (dx, -dy, dz))
    verts.append("%f %f %f" % (dx, dy, dz))
    # Bottom triangle
    verts.append("%f %f %f" % (-dx, 0.0, -dz))
    verts.append("%f %f %f" % (dx, -dy, -dz))
    verts.append("%f %f %f" % (dx, dy, -dz))

    return " ".join(verts)

class Robot:
    def __init__(self):
        self.pos = [ 0.0, 0.0, 0.0 ]
        self.rot = [ 0.0, 0.0, 0.0 ]

        self.dims = [ 0.5, 0.5, 2.0 ]

        self.pose_id = None
        self.pose_wmes = [ SoarWME(d, 0.0) for d in ['x', 'y', 'z', 'roll', 'pitch', 'yaw' ] ]
        self.update_pose = True

        self.self_id = None
        self.arm_id = None

        self.moving_status = SoarWME("moving-status", "stopped")
        self.held_object = SoarWME("holding-object", "none")
        self.current_wp = SoarWME("current-waypoint", "none")

    def get_held_object(self):
        return self.held_object.get_value()
    
    def set_moving_status(self, status):
        self.moving_status.set_value(status)

    def update(self, robot_info):
        self.held_object.set_value("none" if robot_info.held_object is None else str(robot_info.held_object))
        self.current_wp.set_value(robot_info.current_waypoint)

        for d in range(3):
            if abs(self.pos[d] - robot_info.xyzrpy[d]) > 0.02:
                self.pos[d] = robot_info.xyzrpy[d]
                self.pose_wmes[d].set_value(robot_info.xyzrpy[d])
                self.update_pose = True

            if abs(self.rot[d] - robot_info.xyzrpy[3+d]) > 0.05:
                self.rot[d] = robot_info.xyzrpy[3+d]
                self.pose_wmes[3+d].set_value(robot_info.xyzrpy[3+d])
                self.update_pose = True
    
    ############# Soar Interface Functions #############

    def is_added(self):
        return self.self_id is not None

    def add_to_wm(self, parent_id, svs_commands):
        if self.self_id is not None:
            self.update_wm(parent_id, svs_commands)
            return

        self.self_id = parent_id.CreateIdWME("self")
        self.moving_status.add_to_wm(self.self_id)
        self.current_wp.add_to_wm(self.self_id)

        self.pose_id = self.self_id.CreateIdWME("pose")
        for wme in self.pose_wmes:
            wme.add_to_wm(self.pose_id)

        self.arm_id = self.self_id.CreateIdWME("arm")
        self.arm_id.CreateStringWME("moving-status", "wait")
        self.held_object.add_to_wm(self.arm_id)

        svs_commands.append(SVSCommands.add_node("robot", parent="world", pos=self.pos, rot=self.rot))
        svs_commands.append(SVSCommands.add_node("robot_pos", parent="robot"))
        svs_commands.append(SVSCommands.add_box("robot_body", parent="robot", pos=[0.2, 0.0, 1.0], scl=self.dims))
        svs_commands.append("add robot_view robot v %s p %f %f %f" % \
                ( get_view_region_verts(), VIEW_DIST/2 + .5, 0.0, VIEW_HEIGHT/2-self.pos[2] - 0.1))
    
    def update_wm(self, parent_id, svs_commands):
        if self.self_id is None:
            self.add_to_wm(parent_id, svs_commands)
            return

        self.moving_status.update_wm()
        self.held_object.update_wm()
        self.current_wp.update_wm()

        if self.update_pose:
            for wme in self.pose_wmes:
                wme.update_wm()
            svs_commands.append(SVSCommands.change_pos("robot", self.pos))
            svs_commands.append(SVSCommands.change_rot("robot", self.rot))
            self.update_pose = False
    
    def remove_from_wm(self, svs_commands):
        if self.self_id is None: return

        self.moving_status.remove_from_wm()
        self.held_object.remove_from_wm()
        self.current_wp.remove_from_wm()

        for wme in self.pose_wmes:
            wme.remove_from_wm()
        
        self.self_id.DestroyWME()
        self.self_id = None
        self.arm_id = None
        self.pose_id = None
        
        svs_commands.append("delete robot")
    

File: mobile-sim/java/src/soargroup/rosie/perception/Robot.java
package soargroup.rosie.perception;

import java.util.HashSet;
import java.util.Properties;

import soargroup.mobilesim.lcmtypes.robot_info_t;
import soargroup.mobilesim.lcmtypes.tag_classification_list_t;
import soargroup.mobilesim.lcmtypes.tag_classification_t;
import sml.Identifier;
import edu.umich.rosie.soar.ISoarObject;
import edu.umich.rosie.soar.SVSCommands;
import edu.umich.rosie.soar.SoarUtil;
import edu.umich.rosie.soar.StringWME;
import edu.umich.rosie.soarobjects.Pose;

public class Robot implements ISoarObject {
	private static final double VIEW_DIST = 8.0;
	private static final double VIEW_ANGLE = Math.PI/2 * .8;
	private static final double VIEW_HEIGHT = 2.0;
	
	private static final double[] dims = new double[]{.5, .5, .5};
	
	private double[] pos = new double[]{ 0.0, 0.0, 0.0 };
	private double[] rot = new double[]{ 0.0, 0.0, 0.0 };
	private Pose pose;
	
	private boolean updatePose = true;
	
	private StringBuilder svsCommands = new StringBuilder();
	
	private Identifier selfId = null;
	private Identifier armId = null;
	
	private StringWME movingState;
	private StringWME heldObject;
	
	private StringWME waypointId;
	
	public Robot(Properties props){
		movingState = new StringWME("moving-status", "stopped");
		heldObject = new StringWME("holding-object", "none");
		waypointId = new StringWME("current-waypoint", "none");
		pose = new Pose();
	}
	
	public String getHeldObject(){
		return this.heldObject.getValue();
	}
	
	public void setHeldObject(String handle){
		this.heldObject.setValue(handle == null ? "none" : handle);
	}
	
	public void setPos(double[] xy){
		pos[0] = xy[0];
		pos[1] = xy[1];
		pose.setX(pos[0]);
		pose.setY(pos[1]);
		updatePose = true;
	}

  public double[] getXyzrpy(){
    double[] xyzrpy = new double[]{
      pos[0], pos[1], pos[2], rot[0], rot[1], rot[2]
    };
    return xyzrpy;
  }
	
	public void update(robot_info_t info){
		for(int d = 0; d < 3; d++){
			if(Math.abs(pos[d] - info.xyzrpy[d]) > 0.02){
				pos[d] = info.xyzrpy[d];
				updatePose = true;
			}
		}
		for(int d = 0; d < 3; d++){
			if(Math.abs(rot[d] - info.xyzrpy[3+d]) > 0.05){
				rot[d] = info.xyzrpy[3+d];
				updatePose = true;
			}
		}
		if(updatePose){
			pose.updateWithArray(info.xyzrpy);
		}
	}
	
	 public void updateMovingState(String newMovingState){
		 movingState.setValue(newMovingState);
	 }
	
	/* Creates a triangular view region of height VIEW_DIST
	 * and angle VIEW_ANGLE and a height of 2m
	 */
	public static String getViewRegionVertices(){
		StringBuilder verts = new StringBuilder();
		double dx = VIEW_DIST/2;
		double dy = VIEW_DIST * Math.sin(VIEW_ANGLE/2);
		double dz = VIEW_HEIGHT/2;
		// Top triangle
		verts.append(String.format("%f %f %f ", -dx, 0.0, dz));
		verts.append(String.format("%f %f %f ", dx, -dy, dz));
		verts.append(String.format("%f %f %f ", dx, dy, dz));
		// Top triangle
		verts.append(String.format("%f %f %f ", -dx, 0.0, -dz));
		verts.append(String.format("%f %f %f ", dx, -dy, -dz));
		verts.append(String.format("%f %f %f", dx, dy, -dz));
		
		return verts.toString();
	}
	
	public String getSVSCommands(){
		String cmds = svsCommands.toString();
		svsCommands = new StringBuilder();
		return cmds;
	}
	
	private boolean added = false;

	public boolean isAdded() {
		return added;
	}

	@Override
	public void addToWM(Identifier parentId) {
		if(added){
			return;
		}
		selfId = parentId.CreateIdWME("self");
		pose.addToWM(selfId);
		movingState.addToWM(selfId);

		armId = selfId.CreateIdWME("arm");
		armId.CreateStringWME("moving-status", "wait");
		heldObject.addToWM(armId);
		
		svsCommands.append(String.format("add robot world p %s r %s\n", 
				SVSCommands.posToStr(pos), SVSCommands.rotToStr(rot)));
		svsCommands.append(String.format("add robot_pos robot\n"));
		svsCommands.append(String.format("add robot_body robot v %s p .2 0 0 s %s\n", 
				SVSCommands.bboxVertices(), SVSCommands.scaleToStr(dims)));
		svsCommands.append(String.format("add robot_view robot v %s p %f %f %f\n", 
				getViewRegionVertices(), VIEW_DIST/2 + .5, 0.0, VIEW_HEIGHT/2 - dims[2]/2));

		added = true;
	}

	@Override
	public void updateWM() {
		if(!added){
			return;
		}

		movingState.updateWM();
		heldObject.updateWM();
		waypointId.updateWM();

		if(updatePose){
			pose.updateWM();
			svsCommands.append(SVSCommands.changePos("robot", pos));
			svsCommands.append(SVSCommands.changeRot("robot", rot));
			updatePose = false;
		}
	}
	
	@Override
	public void removeFromWM() {
		if(!added){
			return;
		}
		
		movingState.removeFromWM();
		heldObject.removeFromWM();
		waypointId.removeFromWM();
		pose.removeFromWM();

		selfId.DestroyWME();
		selfId = null;
		armId = null;
		
		svsCommands.append(String.format("delete robot\n"));
		added = false;
	}

}
